### PR TITLE
fix: inspect laneStatusCoded to deduce direction

### PR DESCRIPTION
**Pull request by hbruch**
This PR fixes an issue with the DATEX2CIFS transformation, where for constructions always BOTH_DIRECTIONS was assumed.

Now, the encoded lanestatus (see [DATEX2 Profil Datenmodell für Baustellen 4-00-00](https://www.mdm-portal.de/wp-content/uploads/2019/03/mdm_datenmodell_baustellen_04-00-00.zip)) is evaluated.:
image
image

A situationRecord is assumed to concern only one direction, if lanes in opposite direction are unaffected, i.e.
* no lane in opposite direction is directed to the right-hand-side carriageway (no `w`, `i`, `u` right to the centre line (`1` or `2`)
* no lane in opposite direction is width restricted or closed (represented by char `i` or `x`) and 
* no lane in direction is directed to the left-hand-side carriageway (  only `s`,`l`, `u` left to the centre line (`1` or `2`))

### CODE REVIEW

File: pipeline/transformer/cifs.py [hunk 1]
```diff
@@ -130,19 +132,62 @@ def _should_skip(self, situation, situationRecord):
...
+        lanesPerDirection = lanestatus.replace('2', '1').split('1')
+
+        leftLanes = lanesPerDirection[0]
+        rightLanes = lanesPerDirection[1]
+
+        # if leftLanes include more than unnrestricted lane, should, shoulder separatore, or
+        # some lanes of opposite directions are switched to the right lines, opposite direction is concerned
+        return len(re.sub('[usl]', '', leftLanes)) > 0 < len(leftLanes) or len(re.sub('[^uiw]', '', rightLanes)) > 0
+
     def _detect_direction(self, situation, situationRecord):
```
Comment (the-infinity) [inline pipeline/transformer/cifs.py:166]: Missing response type

File: pipeline/transformer/cifs.py [hunk 2]
```diff
@@ -130,19 +132,62 @@ def _should_skip(self, situation, situationRecord):
 
         return False
 
+    def _laneStatusCoded(self, situationRecord) -> Optional[str]:
```
Comment (the-infinity) [inline pipeline/transformer/cifs.py:135]: Missing type

File: pipeline/transformer/cifs.py [hunk 3]
```diff
@@ -130,19 +132,62 @@ def _should_skip(self, situation, situationRecord):
...
+        """
```
Comment (the-infinity) [inline pipeline/transformer/cifs.py:136]: Maybe introduce this as "example data"?

File: pipeline/transformer/cifs.py [hunk 4]
```diff
@@ -130,19 +132,62 @@ def _should_skip(self, situation, situationRecord):
...
+        </situationRecord>
+        """
+        lsElement = situationRecord.find('d:impact/d:impactExtension/d:impactExtended/d:laneStatusCoded', ns)
+        return lsElement.text if lsElement is not None else None
+
+    @staticmethod
+    def _is_opposite_direction_concerned(lanestatus: str) -> bool:
+        # lanes can be single carriageways (encoded by centre line '1', or dual carriagways (encoded by '2'))
+        # we split at both.
+        lanesPerDirection = lanestatus.replace('2', '1').split('1')
```
Comment (derhuerst) [inline pipeline/transformer/cifs.py:157]: Can you add tests for this logic? That would make it a lot easier to grasp.

Comment (hbruch) [reply pipeline/transformer/cifs.py:162]: Done with e57911d.

To test time dependant filtering of constructions, I added a new parameter `current_time` so the point in time can be fixed for the test.

File: pipeline/transformer/cifs.py (hunk 3 above)
Comment (hbruch) [reply pipeline/transformer/cifs.py:137]: Done with ff26ebe

File: pipeline/transformer/cifs.py (hunk 2 above)
Comment (hbruch) [reply pipeline/transformer/cifs.py]: Done with https://github.com/mobidata-bw/ipl-dagster-pipeline/commit/ff26ebe953f0da1da41b073f17a9c6dacf236343

File: pipeline/transformer/cifs.py (hunk 1 above)
Comment (hbruch) [reply pipeline/transformer/cifs.py]: Done with https://github.com/mobidata-bw/ipl-dagster-pipeline/commit/ff26ebe953f0da1da41b073f17a9c6dacf236343

File: pipeline/transformer/cifs.py [hunk 5]
```diff
@@ -130,21 +133,68 @@ def _should_skip(self, situation, situationRecord):
 
         return False
 
-    def _detect_direction(self, situation, situationRecord):
+    def _laneStatusCoded(self, situationRecord: ET) -> Optional[str]:
+        """
+        Extracts the laneStatusCoded value of the given situationoRecord or None, if not available.
```
Comment (derhuerst) [inline pipeline/transformer/cifs.py:138]:
```suggestion
        Extracts the laneStatusCoded value of the given situationRecord or None, if not available.
```

File: pipeline/transformer/cifs.py [hunk 6]
```diff
@@ -130,21 +133,68 @@ def _should_skip(self, situation, situationRecord):
...
+                </impactExtension>
+            </impact>
+        </situationRecord>
+        """
+        lsElement = situationRecord.find('d:impact/d:impactExtension/d:impactExtended/d:laneStatusCoded', ns)
+        return lsElement.text if lsElement is not None else None
+
+    @staticmethod
+    def _is_opposite_direction_concerned(lanestatus: str) -> bool:
+        # lanes can be single carriageways (encoded by centre line '1', or dual carriagways (encoded by '2'))
```
Comment (derhuerst) [inline pipeline/transformer/cifs.py:160]: What is meant by "centre line" here? The character in the centre of the `lanestatus` string?

Comment (derhuerst) [inline pipeline/transformer/cifs.py:160]:
```suggestion
        # lanes can be single carriageways (encoded by centre line '1', or dual carriageways (encoded by '2'))
```

File: pipeline/transformer/cifs.py [hunk 7]
```diff
@@ -130,21 +133,68 @@ def _should_skip(self, situation, situationRecord):
...
+        # we split at both.
+        lanesPerDirection = lanestatus.replace('2', '1').split('1')
```
Comment (derhuerst) [inline pipeline/transformer/cifs.py:162]: Is the digit in the centre *guaranteed* to be the only digit in the entire string?

File: pipeline/transformer/cifs.py [hunk 8]
```diff
@@ -130,21 +133,68 @@ def _should_skip(self, situation, situationRecord):
...
+
+        leftLanes = lanesPerDirection[0]
+        rightLanes = lanesPerDirection[1]
+
+        # if leftLanes include more than unnrestricted lane, should, shoulder separatore, or
```
Comment (derhuerst) [inline pipeline/transformer/cifs.py:167]: I don't understand this comment. Can you try to clarify the cases the logic checks for, e.g. by splitting them into individual boolean variables which are then combined into a final check?

File: pipeline/transformer/cifs.py (hunk 6 above)
Comment (hbruch) [reply pipeline/transformer/cifs.py]: fixed via [b3cc60e](https://github.com/mobidata-bw/ipl-dagster-pipeline/pull/108/commits/b3cc60ef65598fc2e49762d98c9b6b61d4a89d44)

Comment (hbruch) [reply pipeline/transformer/cifs.py]: Fixed via [b3cc60e](https://github.com/mobidata-bw/ipl-dagster-pipeline/pull/108/commits/b3cc60ef65598fc2e49762d98c9b6b61d4a89d44)

File: pipeline/transformer/cifs.py (hunk 5 above)
Comment (hbruch) [reply pipeline/transformer/cifs.py]: Fixed via [b3cc60e](https://github.com/mobidata-bw/ipl-dagster-pipeline/pull/108/commits/b3cc60ef65598fc2e49762d98c9b6b61d4a89d44)

File: pipeline/transformer/cifs.py (hunk 7 above)
Comment (hbruch) [reply pipeline/transformer/cifs.py:176]: Now the lanestatus is validated agains a pattern. If it does nont conform, it's ignored. See [0d339f2](https://github.com/mobidata-bw/ipl-dagster-pipeline/pull/108/commits/0d339f2b9037c47f2c42b3e0948fcafcbe313fb3)

File: pipeline/transformer/cifs.py (hunk 8 above)
Comment (hbruch) [reply pipeline/transformer/cifs.py:181]: Split into multiple boolean variables in [e4f36e5](https://github.com/mobidata-bw/ipl-dagster-pipeline/pull/108/commits/e4f36e5de5dc172138a2f7a7e4346dfdb54eede3)

File: pipeline/transformer/cifs.py [hunk 9]
```diff
@@ -130,21 +136,86 @@ def _should_skip(self, situation, situationRecord):
...
+        # some lanes of opposite directions are switched to the right lines, opposite direction is concerned
+        hasAnyLaneLeftToCentreLine = 0 < len(leftLanes)
+        hasLanesOtherThanUnrestrictedAndSideLanes = len(re.sub('[usl]', '', leftLanes)) > 0
```
Comment (derhuerst) [inline pipeline/transformer/cifs.py:184]:
```suggestion
        hasLeftLanesOtherThanUnrestrictedAndSideLanes = len(re.sub('[usl]', '', leftLanes)) > 0
```

**Pull request merged**